rust/src/lib.rs

```diff
@@ -3443,6 +3491,108 @@ pub unsafe extern "C" fn zcashlc_tor_lwd_conn_get_tree_state(
+    // SAFETY: We ensure unwind safety by:
+    // - using `*mut tor::LwdConn` and respecting mutability rules on the Swift side, to
+    //   avoid observing the effects of a panic in another thread.
+    // - discarding the `tor::LwdConn` whenever we get an error that is due to a panic.
+    let lwd_conn = AssertUnwindSafe(lwd_conn);
+
+    let res = catch_panic(|| {
+        let network = parse_network(network_id)?;
+        let mut db_data = unsafe { wallet_db(db_data, db_data_len, network)? };
+        let account_uuid = account_uuid_from_bytes(account_uuid_bytes)?;
+
+        let lwd_conn = unsafe { lwd_conn.as_mut() }
+            .ok_or_else(|| anyhow!("A Tor lightwalletd connection is required"))?;
+
+        // one day's worth of blocks.
+        let exposure_depth = (24 * 60 * 60) / 75;
+        let addrs =
+            db_data.get_ephemeral_transparent_receivers(account_uuid, exposure_depth, true)?;
+
+        // pick the address with the minimum check time that is less than or equal to now (or
+        // absent)
+        let now = SystemTime::now();
+        let selected_addr_meta = addrs
+            .into_iter()
+            .filter(|(_, meta)| {
+                meta.as_ref()
+                    .and_then(|m| m.next_check_time())
+                    .iter()
+                    .all(|t| t <= &now)
+            })
+            .min_by_key(|(_, meta)| meta.as_ref().map(|m| m.next_check_time()));
+
+        let cur_height = db_data
+            .chain_height()?
+            .ok_or(SqliteClientError::ChainHeightUnknown)?;
+
+        let mut found = false;
+        if let Some((addr, meta)) = selected_addr_meta {
+            let addr_str = addr.encode(&network);
+            lwd_conn.with_address_utxo_stream(addr_str, |raw_tx| {
+                found = true;
+                let mined_height = (raw_tx.height > 0).then(|| {
+                    BlockHeight::from(
+                        u32::try_from(raw_tx.height).expect("block heights fit in u32"),
+                    )
+                });
+                let consensus_branch_id =
+                    BranchId::for_height(&network, mined_height.unwrap_or(cur_height + 1));
+
+                let tx = Transaction::read(&raw_tx.data[..], consensus_branch_id)?;
+                decrypt_and_store_transaction(&network, &mut db_data, &tx, mined_height)?;
+
+                Ok(())
+            })?;
+
+            if !found {
+                let blocks_since_exposure = meta
+                    .and_then(|m| match m.exposed_at() {
+                        ExposedAt::Height(block_height) => Some(cur_height - block_height),
+                        ExposedAt::Unexposed => None,
+                        ExposedAt::Unknown => None,
+                    })
+                    .unwrap_or(1);
+
+                // the log_2 of the number of blocks since exposure will always fit in a u32
+                let offset_seconds = f64::from(blocks_since_exposure).log2().round() as u32;
```

Comment:
As noted in pairing, this is meant to be:
```suggestion
                let offset_seconds = f64::from(blocks_since_exposure * 75).log2().round() as u32;
```